Map Modifiers - update

MatejNota opened the pull request:
Hello, I am writing update of map modifiers. I implemented weighted random so now there are weights for each map modifier. I also added option for no effect. 

However, for actual modifiers in maps, i did not implement any modifiers yet. That's because I would like to know, if there should be effect for all maps. Or maybe if there should be several effects to choose from with weighted random again. 

I know this is not perfect, but I am happy to fix any mistakes.

MatejAutik committed:
code changes

MatejAutik committed:
Modifiers with weighted random (currently no map check)

Zenoo reviewed (changes requested):
Please remove your modifications in the `prisma` and `server/prisma/migrations` folders.
You only have a migration adding a field, and another removing it, so no need for any of them.

As for the map modifiers, they should be tied to existing backgrounds.
Right now the background for a fight is chosen with the following:

![image](https://github.com/user-attachments/assets/ec70ba57-9c0b-452d-a106-8d5576a58e59)

You have to add your modifier data directly in the fightBackgrounds, so they are tied to each map.

![image](https://github.com/user-attachments/assets/f3c3cb1b-a4a9-46f1-91c7-c24f8b1c822e)

MatejNota commented:
Ok, I will change it. But I would like to know, if the effects are to be always active. For example:

Let's say selected map is SWAMP: 
Is effect always same? I mean can it differ like one time it would be -50% speed, other time -25% etc

Additionally, which maps should have effect? Even boss / tournament background?

Thank you for explanation.

MatejAutik committed:
Fix of pull request

MatejNota commented:
I did this quick change. I did just one map with static effect. It it like you wanted?

Zenoo reviewed (changes requested):
Yes, this is how it should globally work.
However, instead of a background.effect being a string, it should be a value from an enum declared above, containing every effect.

For example:
```ts
enum MapEffect {
	HalfSpeed
}

...

{ name: '10.png', odds: 1, effect: MapEffect.HalfSpeed },
```

Also, yes the effects should be constant for each map.
But we would need to find many other maps in order to have more variety in both tournament and boss maps.

MatejAutik committed:
Enum class

MatejNota commented:
I added effect for swamp, and effect for dessert. I lowered odds on those maps. 
But I dont know how many effects should I do.

Zenoo commented on server/src/utils/fight/generateFight.ts:
```diff
@@ -147,6 +148,7 @@ const generateFight = async ({
     team2: { brutes: team2.brutes ?? [], backups: team2Backups, bosses: team2.bosses ?? [] },
     modifiers,
     clanFight: clanWar,
+    background: background.effect,
```
```suggestion
    mapEffect: background.effect,
```

Zenoo commented on server/src/utils/fight/getFighters.ts:
```diff
@@ -5,12 +5,14 @@ import {
 import {
   Brute, FightModifier, PrismaClient, SkillName,
 } from '@labrute/prisma';
+import { Console } from 'node:console';
```
```suggestion
```

Zenoo commented on server/src/utils/fight/getFighters.ts:
```diff
@@ -93,6 +95,22 @@ const handleSkills = (brute: Brute, fighter: DetailedFighter) => {
+) => { 
+  switch(effect){
+    case MapEffect.None:
+      break;
```
```suggestion
```

Zenoo commented on server/src/utils/fight/getFighters.ts:
```diff
@@ -93,6 +95,22 @@ const handleSkills = (brute: Brute, fighter: DetailedFighter) => {
+    case MapEffect.HalfSpeed:
+      brute.speedValue *= 0.5;
+    case MapEffect.OneHp:
+      brute.hp *= 0.1;
```
The modifier OneHp is wrongly named, as it doesn't make the fighters have one HP, but a tenth.
Also, HPs should always be integers, so you need to round the result.
Same for the speed, it needs to be rounded

Zenoo commented on server/src/utils/fight/getFighters.ts:
```diff
@@ -130,13 +148,15 @@ type GetFightersParams = {
   team2: Team,
   modifiers: FightModifier[],
   clanFight?: boolean,
+  background : MapEffect,
```
```suggestion
  mapEffect?: MapEffect,
```

Zenoo commented on server/src/utils/fight/getFighters.ts:
```diff
@@ -130,13 +148,15 @@ type GetFightersParams = {
   team2,
   modifiers,
   clanFight,
+  background,
```
```suggestion
  mapEffect,
```

Zenoo commented on server/src/utils/fight/getFighters.ts:
```diff
@@ -168,6 +188,7 @@ const getFighters = ({
       const bruteAgility = getFinalStat(brute, 'agility', modifiers);
 
       handleModifiers(brute, modifiers);
+      handleMapModifiers(brute, background);
```
```suggestion
      handleMapModifiers(brute, mapEffect);
```

Zenoo reviewed (changes requested):
Instead of having MapEffect.None, just make the effect optional and remove .None.

Also, please revert the changes in the prisma folder, as there was no change to the DB.

After the other reviews are fixed, you will also need to display the modifiers somewhere on the fight view, to let people know they were active during that fight

MatejNota committed:
Merge branch 'Zenoo:main' into main

MatejAutik committed:
fixes

MatejAutik committed:
Merge branch 'main' of https://github.com/MatejNota/labrute_fork

MatejAutik committed:
Fixed prisma folder (I think)

MatejNota commented:
I think all should be fixed. If not, i will fix it again. Also about frontend, I have very little experience with it, but I will try my best. 

For making these modifiers visible, I was thinking about making some little squares (with some image) under each fighters portrait. Upon hover there would be description of current map effect. If thre is no map effect, there would be nothing displayed.

Zenoo commented on server/src/utils/fight/getFighters.ts:
```diff
@@ -93,6 +94,20 @@ const handleSkills = (brute: Brute, fighter: DetailedFighter) => {
+) => { 
+  switch(effect){
+    case MapEffect.HalfSpeed:
+      Math.round(brute.speedValue *= 0.5);
```
```suggestion
      brute.speedValue = Math.round(brute.speedValue * 0.5);
```

Zenoo commented on core/src/fight/backgrounds.ts:
```diff
@@ -1,30 +1,36 @@
+export enum MapEffect {
+	HalfSpeed,
+  OneHp,
```
I don't like this modifier, fighters having only one HP doesn't make good fights.

MatejNota replied:
OK, I will remove it.

Zenoo reviewed (changes requested):
Please revert the changes in the prisma folder.

MatejNota commented:
To which project version should i revert prisma folder? When I looked yesterday I only found 2 files which had differences, so i manually fixed it. But I can also just revert, i just don't know to which version.

Zenoo commented:
You can simply remove your changes from the 2 files in the prisma folder.

MatejNota committed:
Merge branch 'Zenoo:main' into main

MatejAutik committed:
removed 1hp modifier

MatejNota commented:
I can't quite figure out where should I code FE. I was looking with inspect element and found that I should probably be altering canvas output (in fight scene). But I can't seem to find where exactly is that in FightView.tsx. 

Can point me in right direction?

Zenoo commented:
No, the canvas shouldn't be altered for this, in my opinion the modifiers should be displayed above or below the fight canvas.

MatejAutik committed:
FE prototype

MatejAutik committed:
sprite for icon

MatejNota commented:
Hello, so far I tried this prototype of front end for map modifier. I put it under fight canvas. Currently it is statically shows, can only be changed if you modify   const currentMapEffect = 0 + 1;

Problem is, I can't quite figure out how to pass background.effect into this file. I tried to import, then export getFighters in core/src/index.ts but I didn't manage to do it. 

But I think when I do it, it will be quite allright display for this modifier.

Zenoo commented on core/src/fight/backgrounds.ts:
```diff
@@ -1,30 +1,35 @@
+  { name: '9.png', odds: 2},
+  { name: '10.png', odds: 1, effect: MapEffect.HalfSpeed},
+  { name: '11.png', odds: 2},
+  { name: '12.png', odds: 1},
```
No need to lower the odds if no modifier

```suggestion
  { name: '12.png', odds: 2},
```

Zenoo commented on client/src/views/FightView.tsx:
```diff
@@ -52,6 +52,8 @@ const FightView = () => {
     return [firstAd, secondAd];
   }, [language]);
 
+  const currentMapEffect = 0 + 1;
```
You can check which background is used in the fight here instead, you have access to the fight variable

Zenoo reviewed (changes requested):
The files modified in prisma are still here, please just revert their modifications ...

Also, please add a screenshot of what your displayed modifier looks like in this PR

MatejNota commented:
![map_modifier](https://github.com/user-attachments/assets/e244d8a0-58eb-4abe-a139-b1d4b007365c)
In this screenshot I have mouse on icon, so there is active hover for explanation of effect

MatejAutik committed:
FE + fixes in files

MatejNota commented:
What I did:

- I fixed odds in backgrounds.ts
- I moved up call for handleMapModifiers in getFighters.ts, because value changed, but it was wrongly displayed
- I reverted mapEffect: (background as FightBackground).effect, in generateFights.ts, because in mapEffect: background.effect, I get error that property effect doesn't exist on type '{ name: string; odds: number; }'.
- I changed condition for displaying of modifiers in fight

Now I will look into the problem with folder prisma, because at first glance I did not find any different files. Also, for clarification, do you mean prisma, or server/prisma? Because I had some problems with migrations that I fixed locally. (There was some king of migration that dropped "lastSeen" and some other things)

Zenoo commented:
You can easily check what changed in your PR files here:
https://github.com/Zenoo/labrute/pull/1050/files

It's in the prisma folder at the project root

Zenoo commented on client/src/views/FightView.tsx:
```diff
@@ -89,8 +91,26 @@ const FightView = () => {
+          <Box sx={{ ml: 5, alignSelf: 'center', position: 'relative' }}>
+            {/* Fight Component */}
             <FightComponent fight={fight} />
+            {currentMapEffect === '10.png' && (
```
Instead of checking for a single background, you have to find the current background object from the fightBackgrounds list, and find its effect.
If there is none, display nothing, if there is one, select the correct thing to display

MatejNota replied:
I looked into this, and only solution I see is to alter model Fight in database to also hold .effect. Or maybe it is possible to import out of scope of /core but I don't see a way how to do it

Zenoo commented on core/src/fight/backgrounds.ts:
```diff
@@ -1,30 +1,35 @@
+  effect?: MapEffect,
 };
 
 export const tournamentBackground = {
```
```suggestion
export const tournamentBackground: FightBackground = {
```

Zenoo commented on core/src/fight/backgrounds.ts:
```diff
@@ -1,30 +1,35 @@
+  odds: 0
 };
 
 export const bossBackground = {
```
```suggestion
export const bossBackground: FightBackground = {
```

Zenoo commented on server/src/utils/fight/generateFight.ts:
```diff
@@ -147,6 +147,7 @@ const generateFight = async ({
     team2: { brutes: team2.brutes ?? [], backups: team2Backups, bosses: team2.bosses ?? [] },
     modifiers,
     clanFight: clanWar,
+    mapEffect: (background as FightBackground).effect,
```
After typing the boss and tournament backgrounds correctly, you don't need to force type here

```suggestion
    mapEffect: background.effect,
```

MatejNota replied:
Oh I get it now

Zenoo reviewed (changes requested):
As for the icon, I think it would be better to display it here, small enough to fix above the titles, in absolute

![image](https://github.com/user-attachments/assets/291d3df5-8271-41d7-b3dd-bbae45890d32)

MatejAutik committed:
UI changes + added empty line in edge.js

MatejNota commented:
I moved the icon to the top, but I have not changed the logic behind map effects (as I wrote in my reply to your comment). 

Regarding the prisma file changes, I added extra line in edge.js, it seems it resolved the diff issue for that file. But in index.js, I can't seem to resolve diff in files. Even when I copied whole file from main branch it doesn't work, because the file is identical with one on my fork and thus I am unable to commit and push it.

MatejNota commented:
![moved_icon](https://github.com/user-attachments/assets/7b9bbeb0-a691-4fc8-b67e-d5fa0177303e)
This is how the icon looks like now

Zenoo commented:
Are you still working on this @MatejNota ?

MatejNota commented:
Hello, unfortunately I will not continue working on this issue.
